sdk/vppagent/dataplane/chain.go

```diff
@@ -0,0 +1,39 @@
+package dataplane
+
+import (
+	"context"
+
+	"github.com/networkservicemesh/networkservicemesh/dataplane/pkg/apis/dataplane"
+)
+
+type chainKeyType string
+
+const (
+	chainKey chainKeyType = "chain"
+)
+
+type chain struct {
+	handlers []dataplane.DataplaneServer
+	index    int
+}
+
+func (c *chain) Next() dataplane.DataplaneServer {
+	if c.index+1 >= len(c.handlers) {
+		return nil
+	}
+	c.index++
+	return c.handlers[c.index]
+}
+
+// WithChain retruns context with chain of dataplane server handlers
+func WithChain(ctx context.Context, next []dataplane.DataplaneServer) context.Context {
```

Comment:
The Chain is not associated to the ctx, its associated to the Chain object.

The reason for this is that if you do associate the Chain with the ctx, you can't nest them easily.  And building up nested pieces is potentially quite valuable.  Go see: https://github.com/networkservicemesh/networkservicemesh/blob/master/sdk/endpoint/composite.go#L57 for an example.

Reply:
Please more all the With* to context.go to match SDK.

Reply:
thanks, fixed